spec.html

```diff
@@ -32215,26 +32242,88 @@ <h1>
+            1. Return « the Time Zone Identifier Record { [[Identifier]]: *"UTC"*, [[PrimaryIdentifier]]: *"UTC"* } ».
+          1. Let _identifiers_ be the List of String values representing time zones supported by the implementation.
+          1. [declared="comparefn"] Sort _identifiers_ into the same order as if an Array of the same values had been sorted using %Array.prototype.sort% with *undefined* as _comparefn_.
+          1. Let _result_ be a new empty List.
+          1. For each element _identifier_ of _identifiers_, do
+            1. Let _primary_ be _identifier_.
+            1. If _identifier_ is a non-primary time zone identifier in this implementation and _identifier_ is not *"UTC"*, then
+              1. Set _primary_ to the name of the primary time zone identifier that _identifier_ resolves to in this implementation.
+              1. NOTE: If _identifier_ resolves to another non-primary time zone identifier, then the implementation must continue resolution of the entire chain until its terminal primary time zone identifier.
+            1. Let _record_ be the Time Zone Identifier Record { [[Identifier]]: _identifier_, [[PrimaryIdentifier]]: _primary_ }.
+            1. Append _record_ to the end of _result_.
+          1. Assert: One element of _result_ is the Time Zone Identifier Record { [[Identifier]]: *"UTC"*, [[PrimaryIdentifier]]: *"UTC"* }.
+          1. Return _result_.
+        </emu-alg>
+      </emu-clause>
 
-        <p>An ECMAScript implementation that includes the ECMA-402 Internationalization API must implement the DefaultTimeZone abstract operation as specified in the ECMA-402 specification.</p>
-        <p>The default implementation of DefaultTimeZone, to be used for ECMAScript implementations that do not include local political rules for any time zones, performs the following steps when called:</p>
+      <emu-clause id="sec-systemtimezoneidentifier" oldids="sec-defaulttimezone" type="implementation-defined abstract operation">
+        <h1>SystemTimeZoneIdentifier ( ): a String</h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>
+            It returns a String value representing the environment's current time zone, which is either a String representing a UTC offset for which IsTimeZoneOffsetString returns *true*, or a primary time zone identifier.
+          </dd>
+        </dl>
 
         <emu-alg>
-          1. Return *"UTC"*.
+          1. If the implementation only supports the UTC time zone, return *"UTC"*.
+          1. Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either a named time zone identifier or a UTC offset string.
+          1. If IsTimeZoneOffsetString(_systemTimeZoneString_) is *true*, return _systemTimeZoneString_.
+          1. Assert: _systemTimeZoneString_ is the value of the [[PrimaryIdentifier]] field of at least one Time Zone Identifier Record returned by AvailableNamedTimeZoneIdentifiers().
```

Comment:
This also seems to impose a substitution requirement, right? Once AvailableNamedTimeZoneIdentifiers has locked in a collection of named time zones, a host report of a time zone identifier outside that collection could not be returned from SystemTimeZoneIdentifier (e.g., observable absence of Etc/Limbo could not be followed by observable presence if the host starts reporting that as the system time zone).

Reply:
It wouldn't and in fact must not, but that is complicated by the fact that a host can change what it reports to an ECMAScript implementation at any time (with observable consequences, e.g. invoke `String(new Date())` in the same agent both before and after changing your system time zone).

Reply:
Is there a system where you could observe in JS one list of timezones, and then change your timezone to a new one that wasn't in the old list, and observe in the same JS program a timezone not in the original list?

Reply:
I'm not sure what you mean, but [Intl.supportedValuesOf](https://tc39.es/ecma402/#sec-intl.supportedvaluesof) exposes the list of known time zones (each of which must be recognized as valid for input) and `new Intl.DateTimeFormat().resolvedOptions().timeZone` exposes the current time zone (which must also be recognized as valid for input). But an ECMAScript implementation has no control over its host, and a naïve implementation might blindly report a value from an unusually-reconfigured host environment that ECMAScript code has already observed to be invalid.

Reply:
Right - I'm asking if there's an actual host out there that can add a new timezone without a restart, such that JS could observe a timezone that wasn't in the original `supportedValuesOf`? iow, is this an actual or a (still important, but) theoretical concern?

Reply:
I'm pretty sure most implementations use different paths for getting the list of timezones (which are generally compiled into their artefacts at build time) vs. getting the current time zone (which generally comes from a syscall to the host environment that can return arbitrary and unstable values), but they interact in the requirement that the current time zone exposed to ECMAScript code must be accepted as valid _from_ ECMAScript code. So this is an actual concern that occurs any time the host environment reports a novel time zone to the ECMAScript implementation, with the probability for a particular implementation to encounter such circumstances increasing with its age.

For example, Europe/Kyiv was added to the IANA TZDB in August 2022 with release 2022b and Asia/Chita was added in release 2014f (the former supplanting the preexisting Europe/Kiev with no data changes, but the latter being created as an entirely new zone), and an implementation with a baked-in list that does not include such a zone can still encounter it from a host with its own more recent TZDB that gets configured to use that zone while the ECMAScript implementation is already running an agent.

Reply:
As noted [above](https://github.com/tc39/ecma262/pull/3035#discussion_r1204634727), canonicalization of the system time zone ID is required by ECMA-402, so diverging 262 and 402 would be out of scope for an editorial PR.

Also, AFAIK the agent-lifetime stability requirements are already (albeit confusingly worded) in the spec today, and presumably changing them would be normative. So I'd prefer to keep the current text as-is, and if we need to change the stability guarantees in the future then do that in a separate normative PR.  Would that be OK?

> if there's an actual host out there that can add a new timezone without a restart, such that JS could observe a timezone that wasn't in the original supportedValuesOf? iow, is this an actual or a (still important, but) theoretical concern?

Windows can apply time zone changes without a reboot. My understanding is that Linux can too. Perhaps Android and iOS too. That said, my understanding is that all current timezone-aware implementations go through ICU and CLDR to deal with time zone identifiers, and ICU code is baked into the implementation. So I find it very unlikely that any current implementation that uses time zones can update ICU without restarting existing agents. So I think it's a theoretical concern today.

Reply:
I spoke with @gibson042 offline and we'll resolve this thread as out-of-scope to this editorial PR, and will continue the discussion in proposal-canonical-tz and with TG2.